cmd/backup-manager/app/util/util.go

```diff
@@ -161,70 +161,109 @@ func GetOptionValueFromEnv(option, envPrefix string) string {
+		}
 	}
+
 	return args, nil
 }
 
 // ConstructDumplingOptionsForBackup constructs dumpling options for backup
 func ConstructDumplingOptionsForBackup(backup *v1alpha1.Backup) []string {
 	var args []string
-	config := backup.Spec.Dumpling
-	if config == nil {
+	config := backup.Spec
+
+	if config.TableFilter != nil && len(config.TableFilter) > 0 {
+		for _, tableFilter := range config.TableFilter {
+			args = append(args, "--filter", tableFilter)
+		}
+	} else {
+		if config.Dumpling != nil && config.Dumpling.TableFilter != nil && len(config.Dumpling.TableFilter) > 0 {
+			for _, tableFilter := range config.Dumpling.TableFilter {
+				args = append(args, "--filter", tableFilter)
+			}
+		} else {
+			args = append(args, defaultTableFilterOptions...)
+		}
+	}
+
+	if config.Dumpling == nil {
 		args = append(args, defaultOptions...)
-		args = append(args, defaultTableFilterOptions...)
 		return args
 	}
 
-	if len(config.Options) != 0 {
-		args = append(args, config.Options...)
+	if len(config.Dumpling.Options) != 0 {
+		args = append(args, config.Dumpling.Options...)
 	} else {
 		args = append(args, defaultOptions...)
 	}
 
-	if len(config.TableFilter) > 0 {
-		for _, tableFilter := range config.TableFilter {
-			args = append(args, "--filter", tableFilter)
-		}
-	} else {
-		args = append(args, defaultTableFilterOptions...)
-	}
 	return args
 }
 
 // ConstructBRGlobalOptionsForRestore constructs BR global options for restore.
 func ConstructBRGlobalOptionsForRestore(restore *v1alpha1.Restore) ([]string, error) {
 	var args []string
-	config := restore.Spec.BR
-	if config == nil {
+	config := restore.Spec
+	if config.BR == nil {
 		return nil, fmt.Errorf("no config for br in restore %s/%s", restore.Namespace, restore.Name)
 	}
-	args = append(args, constructBRGlobalOptions(config)...)
+	args = append(args, constructBRGlobalOptions(config.BR)...)
 	storageArgs, err := getRemoteStorage(restore.Spec.StorageProvider)
 	if err != nil {
 		return nil, err
 	}
 	args = append(args, storageArgs...)
-	if (restore.Spec.Type == v1alpha1.BackupTypeDB || restore.Spec.Type == v1alpha1.BackupTypeTable) && config.DB != "" {
-		args = append(args, fmt.Sprintf("--db=%s", config.DB))
+
+	if config.TableFilter != nil && len(config.TableFilter) > 0 {
+		for _, tableFilter := range config.TableFilter {
+			args = append(args, "--filter", tableFilter)
+		}
+		return args, nil
 	}
-	if restore.Spec.Type == v1alpha1.BackupTypeTable && config.Table != "" {
-		args = append(args, fmt.Sprintf("--table=%s", config.Table))
+
+	switch restore.Spec.Type {
+	case v1alpha1.BackupTypeTable:
+		if config.BR.Table != "" {
+			args = append(args, fmt.Sprintf("--table=%s", config.BR.Table))
+		}
+		if config.BR.DB != "" {
+			args = append(args, fmt.Sprintf("--db=%s", config.BR.DB))
+		}
+	case v1alpha1.BackupTypeDB:
+		if config.BR.DB != "" {
+			args = append(args, fmt.Sprintf("--db=%s", config.BR.DB))
+		}
 	}
+
 	return args, nil
 }
```

Comment:
Can we move the L240-268 into function `constructBRGlobalOptions`? seems this part can be reused(duplicated with L169-L196).

Reply:
Makes sense. WIll have a look at it early next week. I've done a test for BR but it's not passing yet. Will get it done then too.

Reply:
Sorry it's taken me so long but I've finally got some time to work on this again. I've pushed a test for the brbackup. If that's ok then I can easily do the restore one as it's very similar. I've looked at refactoring this like you suggest @lonng. Both these are working with different types though. One is a *v1alpha1.Backup and the other is a *v1alpha1.Restore. I can't really see a clean way of pulling this out `constructBRGlobalOptions`.

Reply:
Sorry for the delay on this. I've added the br backup test. If this is ok I can easily do the restore one as it's very similar. @lonng I've looked at refactoring it like you suggest and as they work with different types I'm not aware of a clean way to do this. We could pass in the v1alpha1 type and check against that but in my mind this will actually make it less readable as we will have more checks needed. Perhaps you know a better of doing this though.

Reply:
Ignore my suggestion. They have different types.